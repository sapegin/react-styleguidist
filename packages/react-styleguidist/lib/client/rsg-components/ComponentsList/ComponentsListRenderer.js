import "core-js/modules/es6.string.small";

import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import Link from 'rsg-components/Link';
import Styled from 'rsg-components/Styled';

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const styles = function styles(_ref) {
  const color = _ref.color;

      
const fontFamily = _ref.fontFamily;

      
const fontSize = _ref.fontSize;

      
const space = _ref.space;

      
const mq = _ref.mq;
  return {
    list: {
      margin: 0,
      paddingLeft: space[2]
    },
    item: {
      color: color.base,
      display: 'block',
      margin: [[space[1], 0, space[1], 0]],
      fontFamily: fontFamily.base,
      fontSize: fontSize.base,
      listStyle: 'none',
      overflow: 'hidden',
      textOverflow: 'ellipsis'
    },
    isChild: _defineProperty({}, mq.small, {
      display: 'inline-block',
      margin: [[0, space[1], 0, 0]]
    }),
    heading: {
      color: color.base,
      marginTop: space[1],
      fontFamily: fontFamily.base,
      fontWeight: 'bold'
    }
  };
};

export function ComponentsListRenderer(_ref2) {
  const classes = _ref2.classes;

      
let items = _ref2.items;
  items = items.filter(function (item) {
    return item.visibleName;
  });

  if (!items.length) {
    return null;
  }

  return React.createElement("ul", {
    className: classes.list
  }, items.map(function (_ref3) {
    const heading = _ref3.heading;

        
const visibleName = _ref3.visibleName;

        
const href = _ref3.href;

        
const content = _ref3.content;
    return React.createElement("li", {
      className: cx(classes.item, (!content || !content.props.items.length) && classes.isChild),
      key: href
    }, React.createElement(Link, {
      className: cx(heading && classes.heading),
      href
    }, visibleName), content);
  }));
}
ComponentsListRenderer.propTypes = {
  items: PropTypes.array.isRequired,
  classes: PropTypes.object.isRequired
};
export default Styled(styles)(ComponentsListRenderer);