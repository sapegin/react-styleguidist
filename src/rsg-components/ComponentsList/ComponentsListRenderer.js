import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import Link from 'rsg-components/Link';
import Styled from 'rsg-components/Styled';

const styles = ({ font, small }) => ({
	list: {
		margin: 0,
		paddingLeft: 15,
	},
	item: {
		display: 'block',
		margin: [[7, 0, 7, 0]],
		fontFamily: font,
		fontSize: 15,
		listStyle: 'none',
		overflow: 'hidden',
		textOverflow: 'ellipsis',
	},
	isChild: {
		[small]: {
			display: 'inline-block',
			margin: [[0, 7, 0, 0]],
		},
	},
	heading: {
		marginTop: 7,
		fontFamily: font,
		fontWeight: 'bold',
	},
});

export function ComponentsListRenderer({ classes, items }) {
	items = items.filter(item => item.name);

	if (!items.length) {
		return null;
	}

	return (
		<ul className={classes.list}>
			{items.map(({ heading, name, slug, content }) => (
				<li className={cx(classes.item, (!content || !content.props.items.length) && classes.isChild)} key={name}>
					<Link className={cx(heading && classes.heading)} href={`#${slug}`}>{name}</Link>
					{content}
				</li>
			))}
		</ul>
	);
}

ComponentsListRenderer.propTypes = {
	items: PropTypes.array.isRequired,
	classes: PropTypes.object.isRequired,
};

export default Styled(styles)(ComponentsListRenderer);
