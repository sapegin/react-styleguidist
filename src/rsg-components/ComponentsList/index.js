export { default } from 'rsg-components/ComponentsList/ComponentsListRenderer';
