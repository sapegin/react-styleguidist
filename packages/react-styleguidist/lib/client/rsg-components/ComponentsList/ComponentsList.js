import "core-js/modules/web.dom.iterable";
import "core-js/modules/es6.array.iterator";
import "core-js/modules/es6.object.keys";
import "core-js/modules/es6.function.name";

import React from 'react';
import ComponentsListRenderer from 'rsg-components/ComponentsList/ComponentsListRenderer';
import PropTypes from 'prop-types';
import getUrl from '../../utils/getUrl';

function _objectSpread(target) { for (let i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; let ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function ComponentsList(_ref) {
  const classes = _ref.classes;

      
const items = _ref.items;

      
const _ref$useRouterLinks = _ref.useRouterLinks;

      
const useRouterLinks = _ref$useRouterLinks === void 0 ? false : _ref$useRouterLinks;

      
const useHashId = _ref.useHashId;

      
const hashPath = _ref.hashPath;
  const mappedItems = items.map(function (item) {
    return _objectSpread({}, item, {
      href: getUrl({
        name: item.name,
        slug: item.slug,
        anchor: !useRouterLinks,
        hashPath: useRouterLinks ? hashPath : false,
        id: useRouterLinks ? useHashId : false
      })
    });
  });
  return React.createElement(ComponentsListRenderer, {
    classes,
    items: mappedItems
  });
}

ComponentsList.propTypes = {
  items: PropTypes.array.isRequired,
  classes: PropTypes.object,
  hashPath: PropTypes.array,
  useRouterLinks: PropTypes.bool,
  useHashId: PropTypes.bool
};
export default ComponentsList;