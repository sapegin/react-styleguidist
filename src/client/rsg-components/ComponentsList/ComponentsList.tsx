import React from 'react';
import PropTypes from 'prop-types';
import ComponentsListRenderer from 'rsg-components/ComponentsList/ComponentsListRenderer';
import { ComponentViewModel } from 'rsg-components/ReactComponent';
import getUrl from '../../utils/getUrl';

interface ComponentsListProps {
	items: ComponentViewModel[];
	hashPath?: string[];
	useRouterLinks?: boolean;
	useHashId?: boolean;
}

const ComponentsList: React.FunctionComponent<ComponentsListProps> = ({
	items,
	useRouterLinks = false,
	useHashId,
	hashPath,
}) => {
	const mappedItems = items.map(item => ({
		...item,
		shouldOpenInNewTab: !!item.href,
		href: item.href
			? item.href
			: getUrl({
					name: item.name,
					slug: item.slug,
					anchor: !useRouterLinks,
					hashPath: useRouterLinks ? hashPath : false,
					id: useRouterLinks ? useHashId : false,
			  }),
	}));
	return <ComponentsListRenderer items={mappedItems} />;
};

ComponentsList.propTypes = {
	items: PropTypes.array.isRequired,
	hashPath: PropTypes.array,
	useRouterLinks: PropTypes.bool,
	useHashId: PropTypes.bool,
};

export default ComponentsList;
