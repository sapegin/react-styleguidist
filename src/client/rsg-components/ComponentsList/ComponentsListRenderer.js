import React from 'react';
import PropTypes from 'prop-types';
import cx from 'clsx';
import Link from 'rsg-components/Link';
import Styled from 'rsg-components/Styled';
import { useStyleGuideContext } from 'rsg-components/Context';
import { getHash } from '../../utils/handleHash';

const styles = ({ color, fontFamily, fontSize, space, mq }) => ({
	list: {
		margin: 0,
		paddingLeft: space[2],
	},
	item: {
		color: color.base,
		display: 'block',
		margin: [[space[1], 0, space[1], 0]],
		fontFamily: fontFamily.base,
		fontSize: fontSize.base,
		listStyle: 'none',
		overflow: 'hidden',
		textOverflow: 'ellipsis',
	},
	isChild: {
		[mq.small]: {
			display: 'inline-block',
			margin: [[0, space[1], 0, 0]],
		},
	},
	heading: {
		color: color.base,
		marginTop: space[1],
		fontFamily: fontFamily.base,
		fontWeight: 'bold',
	},
	isSelected: {
		fontWeight: 'bold',
	},
});

export function ComponentsListRenderer({ classes, items }) {
	const {
		config: { pagePerSection },
	} = useStyleGuideContext();

	const visibleItems = items.filter(item => item.visibleName);

	if (!visibleItems.length) {
		return null;
	}

	// Match selected component in both basic routing and pagePerSection routing.
	const { hash, pathname } = window.location;
	const windowHash = pathname + (pagePerSection ? hash : getHash(hash));
	return (
		<ul className={classes.list}>
			{visibleItems.map(({ heading, visibleName, href, content, shouldOpenInNewTab }) => {
				const isItemSelected = windowHash === href;
				return (
					<li
						className={cx(classes.item, {
							[classes.isChild]: (!content || !content.props.items.length) && !shouldOpenInNewTab,
							[classes.isSelected]: isItemSelected,
						})}
						key={href}
					>
						<Link
							className={cx(heading && classes.heading)}
							href={href}
							target={shouldOpenInNewTab ? '_blank' : undefined}
						>
							{visibleName}
						</Link>
						{content}
					</li>
				);
			})}
		</ul>
	);
}

ComponentsListRenderer.propTypes = {
	items: PropTypes.array.isRequired,
	classes: PropTypes.object.isRequired,
};

export default Styled(styles)(ComponentsListRenderer);
