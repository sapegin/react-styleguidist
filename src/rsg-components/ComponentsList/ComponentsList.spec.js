import React from 'react';
import ComponentsList from './ComponentsList';
import { ComponentsListRenderer } from './ComponentsListRenderer';

it('should set the correct href for items', () => {
	const components = [
		{
			visibleName: 'Button',
			name: 'Button',
			slug: 'button',
		},
		{
			visibleName: 'Input',
			name: 'Input',
			slug: 'input',
		},
	];

	const actual = shallow(<ComponentsList items={components} classes={{}} />);
	expect(actual).toMatchSnapshot();
});

it('if a custom href is provided, should use it instead of generating internal link', () => {
	const components = [
		{
			visibleName: 'External example',
			name: 'External example',
			href: 'http://example.com',
		},
		{
			visibleName: 'Input',
			name: 'Input',
			slug: 'input',
		},
	];

	const actual = shallow(<ComponentsList items={components} classes={{}} />);
	expect(actual).toMatchSnapshot();
});

it('should set a parameter on link when useHashId is activated', () => {
	const components = [
		{
			visibleName: 'Button',
			name: 'Button',
			slug: 'button',
		},
		{
			visibleName: 'Input',
			name: 'Input',
			slug: 'input',
		},
	];

	const actual = shallow(
		<ComponentsList
			items={components}
			classes={{}}
			useRouterLinks
			hashPath={['Components']}
			useHashId
		/>
	);
	expect(actual).toMatchSnapshot();
});

it('should set a sub route on link when useHashId is deactivated', () => {
	const components = [
		{
			name: 'Button',
			slug: 'button',
		},
		{
			name: 'Input',
			slug: 'input',
		},
	];

	const actual = shallow(
		<ComponentsList
			items={components}
			classes={{}}
			useRouterLinks
			hashPath={['Components']}
			useHashId={false}
		/>
	);
	expect(actual).toMatchSnapshot();
});

it('should render sections with nested components', () => {
	const components = [
		{
			visibleName: 'Button',
			slug: 'button',
			href: '#button',
		},
		{
			visibleName: 'Input',
			slug: 'input',
			href: '#input',
		},
	];
	const actual = shallow(<ComponentsListRenderer items={components} classes={{}} />);

	expect(actual).toMatchSnapshot();
});

it('should return null when the list is empty', () => {
	const actual = shallow(<ComponentsListRenderer items={[]} classes={{}} />);

	expect(actual.getElement()).toBe(null);
});

it('should ignore items without visibleName', () => {
	const components = [
		{
			visibleName: 'Button',
			slug: 'button',
			href: '#button',
		},
		{
			slug: 'input',
			href: '#input',
		},
	];
	const actual = shallow(<ComponentsListRenderer items={components} classes={{}} />);

	expect(actual).toMatchSnapshot();
});
