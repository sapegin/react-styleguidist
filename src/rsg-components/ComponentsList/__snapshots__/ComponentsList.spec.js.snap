// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`if a custom href is provided, should use it instead of generating internal link 1`] = `
<Styled(ComponentsList)
  classes={Object {}}
  items={
    Array [
      Object {
        "href": "http://example.com",
        "name": "External example",
        "visibleName": "External example",
      },
      Object {
        "href": "blank#input",
        "name": "Input",
        "slug": "input",
        "visibleName": "Input",
      },
    ]
  }
/>
`;

exports[`should ignore items without visibleName 1`] = `
<ul>
  <li
    className=""
    key="#button"
  >
    <Styled(Link)
      className=""
      href="#button"
    >
      Button
    </Styled(Link)>
  </li>
</ul>
`;

exports[`should render sections with nested components 1`] = `
<ul>
  <li
    className=""
    key="#button"
  >
    <Styled(Link)
      className=""
      href="#button"
    >
      Button
    </Styled(Link)>
  </li>
  <li
    className=""
    key="#input"
  >
    <Styled(Link)
      className=""
      href="#input"
    >
      Input
    </Styled(Link)>
  </li>
</ul>
`;

exports[`should set a parameter on link when useHashId is activated 1`] = `
<Styled(ComponentsList)
  classes={Object {}}
  items={
    Array [
      Object {
        "href": "blank#/Components?id=button",
        "name": "Button",
        "slug": "button",
        "visibleName": "Button",
      },
      Object {
        "href": "blank#/Components?id=input",
        "name": "Input",
        "slug": "input",
        "visibleName": "Input",
      },
    ]
  }
/>
`;

exports[`should set a sub route on link when useHashId is deactivated 1`] = `
<Styled(ComponentsList)
  classes={Object {}}
  items={
    Array [
      Object {
        "href": "blank#/Components/Button",
        "name": "Button",
        "slug": "button",
      },
      Object {
        "href": "blank#/Components/Input",
        "name": "Input",
        "slug": "input",
      },
    ]
  }
/>
`;

exports[`should set the correct href for items 1`] = `
<Styled(ComponentsList)
  classes={Object {}}
  items={
    Array [
      Object {
        "href": "blank#button",
        "name": "Button",
        "slug": "button",
        "visibleName": "Button",
      },
      Object {
        "href": "blank#input",
        "name": "Input",
        "slug": "input",
        "visibleName": "Input",
      },
    ]
  }
/>
`;
